### ISSUE TITLE
Passing a closure to a template can cause codegen error

### ISSUE THREAD

**Issue by bluenote10** [nim-lang/Nim#3154]
I'm not sure whether my conclusion from the error is correct, but at least this is what I'm observing (latest devel):

    import sequtils

    # it looks like this only happens when
    # nesting a proc in a proc
    proc test() =

      let prefix = "prefix"

      proc closure(s: string): int =
        (prefix & s).len

      let suffixes = @["test"]

      # now these lines are mutually exclusive
      # i.e., I can call one of them, but if I
      # have both, nim will produce faulty C code
      let mapRes = suffixes.mapIt(int, closure(it))
      let single = closure("test")

The problem with the C code is:

    .../src/nimcache/debug.c: In function ‘test_93004’:
    .../src/nimcache/debug.c:504:45: error: ‘HEX3Aenv_93258’ undeclared (first use in this function)
      LOC11.ClPrc = closure_93007; LOC11.ClEnv = HEX3Aenv_93258;
                                             ^
    .../src/nimcache/debug.c:504:45: note: each undeclared identifier is reported only once for each function it appears in

**Comment by bluenote10** [comment]
I'm no longer sure whether it has anything to do with the template. In my production code I get a similar error just by calling a certain closure twice. I cannot reproduce this in this minimized example though.

**Comment by kirbyfan64** [comment]
I'm guessing this is just YALLB (yet another lambda lifting bug).

**Comment by Araq** [comment]
Duplicate of #672.
